instrumentation/spring-amqp/src/main/java/brave/spring/amqp/TracingRabbitListenerAdvice.java

```diff
@@ -0,0 +1,62 @@
+package brave.spring.amqp;
+
+import org.aopalliance.aop.Advice;
+import org.aopalliance.intercept.MethodInterceptor;
+import org.aopalliance.intercept.MethodInvocation;
+import org.springframework.amqp.core.Message;
+import org.springframework.amqp.core.MessageProperties;
+
+import brave.Span;
+import brave.Tracer;
+import brave.Tracer.SpanInScope;
+import brave.Tracing;
+import brave.propagation.Propagation;
+import brave.propagation.TraceContext.Extractor;
+import brave.propagation.TraceContextOrSamplingFlags;
+import org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer;
+
+/**
+ * TracingRabbitListenerAdvice is an AOP advice to be used with the {@link
+ * SimpleMessageListenerContainer#setAdviceChain(Advice...) SimpleMessageListenerContainer} in order
+ * to add tracing functionality to a spring-amqp managed message consumer. While a majority of
+ * brave's instrumentation points is implemented using a delegating wrapper approach, this extension
+ * point was ideal as it covers both programmatic and @RabbitListener annotation driven amqp consumers.
+ */
+public class TracingRabbitListenerAdvice implements MethodInterceptor {
+
+  private static final Propagation.Getter<MessageProperties, String> GETTER =
+      new Propagation.Getter<MessageProperties, String>() {
+
+        @Override public String get(MessageProperties carrier, String key) {
+          return (String) carrier.getHeaders().get(key);
+        }
+      };
+  private final Extractor<MessageProperties> extractor;
+  private final Tracer tracer;
+
+  public TracingRabbitListenerAdvice(Tracing tracing) {
+    this.extractor = tracing.propagation().extractor(GETTER);
+    this.tracer = tracing.tracer();
+  }
+
+  @Override public Object invoke(MethodInvocation methodInvocation) throws Throwable {

```

Comment:
so this will apply to any method invocation, right? Feels like the methods that are consumer in nature (polling a message) would be traced differently than ones that are processor in nature (ex RabbitListeners)

I'll pull the code to try and figure out if this is the case or not

Reply:
Right, so this is only meant to be used with the SimpleMessageListenerContainer (or DirectMessageListenerContainer in 2.0+), and would wrap a span around the *processing* - this is shown in the first screenshot that I'd attached earlier.

If we were to also have a separate span for the actual message consumption (e.g. the 2nd screenshot) that is possible with some changes. The processing would then become a SERVER span

Reply:
Having said that I'd need to dig a bit further into the spring-amqp source as well